### PR TITLE
feat: impl invModPUint512 hint

### LINKED ISSUE #559: `invModPUint512` hint
https://github.com/lambdaclass/cairo-vm/blob/db7fff18c9c1312024ebf4119fcdfeba23657b12/cairo_programs/inv_mod_p_uint512.cairo#L20

https://github.com/starkware-libs/cairo-lang/blob/0e4dab8a6065d80d1c726394f5d9d23cb451706a/src/starkware/starknet/security/whitelists/vrf.json#L56

### PR DESCRIPTION
**Pull request by MartianGreed**
Resolves #559

### CODE REVIEW

**Comment by TAdev0** [comment]
Resolves #559

File: pkg/hintrunner/zero/hintcode.go [hunk 1]
```diff
@@ -57,6 +57,8 @@ const (
...
 	uint256SubCode            string = "def split(num: int, num_bits_shift: int = 128, length: int = 2):\n    a = []\n    for _ in range(length):\n        a.append( num & ((1 << num_bits_shift) - 1) )\n        num = num >> num_bits_shift\n    return tuple(a)\n\ndef pack(z, num_bits_shift: int = 128) -> int:\n    limbs = (z.low, z.high)\n    return sum(limb << (num_bits_shift * i) for i, limb in enumerate(limbs))\n\na = pack(ids.a)\nb = pack(ids.b)\nres = (a - b)%2**256\nres_split = split(res)\nids.res.low = res_split[0]\nids.res.high = res_split[1]"
+	// ------ Uint512 hints related code -------
+	invModPUint512Code string = "from starkware.cairo.common.uint512 import Uint512\n\nvalue = Uint512(ids.value, ids.modulus)\nvalue_inv = value.inv_mod(ids.modulus)\ni"
```
Comment (TAdev0) [inline pkg/hintrunner/zero/hintcode.go:61]: what does this hintcode corresponds to ?

the hint is : 

```
    %{
        def pack_512(u, num_bits_shift: int) -> int:
            limbs = (u.d0, u.d1, u.d2, u.d3)
            return sum(limb << (num_bits_shift * i) for i, limb in enumerate(limbs))

        x = pack_512(ids.x, num_bits_shift = 128)
        p = ids.p.low + (ids.p.high << 128)
        x_inverse_mod_p = pow(x,-1, p)

        x_inverse_mod_p_split = (x_inverse_mod_p & ((1 << 128) - 1), x_inverse_mod_p >> 128)

        ids.x_inverse_mod_p.low = x_inverse_mod_p_split[0]
        ids.x_inverse_mod_p.high = x_inverse_mod_p_split[1]
    %}
```

File: pkg/hintrunner/zero/zerohint_uint512.go [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+	VM "github.com/NethermindEth/cairo-vm-go/pkg/vm"
+	"github.com/consensys/gnark-crypto/ecc/stark-curve/fp"
+)
+
+const (
+	P_LOW  = "201385395114098847380338600778089168199"
+	P_HIGH = "64323764613183177041862057485226039389"
+)
+
+func createInvModPUint512Hinter(resolver hintReferenceResolver) (hinter.Hinter, error) {
```
Comment (TAdev0) [inline pkg/hintrunner/zero/zerohint_uint512.go:16]: can you place the `new..` function before the  `create...` one , and add comments to the `new..` function as done for other hints?

File: pkg/hintrunner/zero/zerohint_uint512.go [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+	}
+
+	return newInvModPUint512Hint(x, xInverseModP), nil
+}
+
+func newInvModPUint512Hint(x, xInverseModP hinter.ResOperander) hinter.Hinter {
+	return &GenericZeroHinter{
+		Name: "InvModPUint512",
+		Op: func(vm *VM.VirtualMachine, _ *hinter.HintRunnerContext) error {
+			// def pack_512(u, num_bits_shift: int) -> int:
```
Comment (TAdev0) [inline pkg/hintrunner/zero/zerohint_uint512.go:34]: comments for hints should be formatted with `//>` instead of `//`

**Review by TAdev0 — CHANGES_REQUESTED** [review]

File: pkg/hintrunner/zero/hintcode.go [hunk 4]
```diff
@@ -57,6 +57,8 @@ const (
...
 	uint256MulDivModCode      string = "a = (ids.a.high << 128) + ids.a.low\nb = (ids.b.high << 128) + ids.b.low\ndiv = (ids.div.high << 128) + ids.div.low\nquotient, remainder = divmod(a * b, div)\n\nids.quotient_low.low = quotient & ((1 << 128) - 1)\nids.quotient_low.high = (quotient >> 128) & ((1 << 128) - 1)\nids.quotient_high.low = (quotient >> 256) & ((1 << 128) - 1)\nids.quotient_high.high = quotient >> 384\nids.remainder.low = remainder & ((1 << 128) - 1)\nids.remainder.high = remainder >> 128"
 	uint256SubCode            string = "def split(num: int, num_bits_shift: int = 128, length: int = 2):\n    a = []\n    for _ in range(length):\n        a.append( num & ((1 << num_bits_shift) - 1) )\n        num = num >> num_bits_shift\n    return tuple(a)\n\ndef pack(z, num_bits_shift: int = 128) -> int:\n    limbs = (z.low, z.high)\n    return sum(limb << (num_bits_shift * i) for i, limb in enumerate(limbs))\n\na = pack(ids.a)\nb = pack(ids.b)\nres = (a - b)%2**256\nres_split = split(res)\nids.res.low = res_split[0]\nids.res.high = res_split[1]"
+	// ------ Uint512 hints related code -------
```
Comment (MaksymMalicki) [inline pkg/hintrunner/zero/hintcode.go:60]: Just to clarify, the whole function uses mostly Uint256 utilities and creates new Uint512 stuct. I think it is reasonable to put this hint code int the uint256 section

File: pkg/hintrunner/zero/zerohint.go [hunk 5]
```diff
@@ -111,6 +111,9 @@ func GetHintFromCode(program *zero.ZeroProgram, rawHint zero.Hint, hintPC uint64
 		return createUint256MulDivModHinter(resolver)
 	case uint256SubCode:
 		return createUint256SubHinter(resolver)
+	// Uint512 hints
+	case invModPUint512Code:
```
Comment (MaksymMalicki) [inline pkg/hintrunner/zero/zerohint.go:115]: same here, I think we can move it to uint256 section.

File: pkg/hintrunner/zero/zerohint_uint512.go [hunk 6]
```diff
@@ -0,0 +1,102 @@
...
+			pack512 := func(lolow, loHigh, hiLow, hiHigh *fp.Element, numBitsShift int) big.Int {
+				var loLowBig, loHighBig, hiLowBig, hiHighBig big.Int
+				lolow.BigInt(&loLowBig)
+				loHigh.BigInt(&loHighBig)
+				hiLow.BigInt(&hiLowBig)
+				hiHigh.BigInt(&hiHighBig)
+
+				return *new(big.Int).Add(new(big.Int).Lsh(&hiHighBig, uint(numBitsShift)), &loLowBig).Add(new(big.Int).Lsh(&hiLowBig, uint(numBitsShift)), &loHighBig)
+			}
+			pack := func(low, high *fp.Element, numBitsShift int) big.Int {
```
Comment (MaksymMalicki) [inline pkg/hintrunner/zero/zerohint_uint512.go:41]: do we need a separate function for that? I would just put this code right above the packing of the two variables, and not make an anonymous function out of that

**Comment by MaksymMalicki** [comment]
Hey, could you add [this integration test](https://github.com/lambdaclass/cairo-vm/blob/db7fff18c9c1312024ebf4119fcdfeba23657b12/cairo_programs/inv_mod_p_uint512.cairo#L20) to the integration tests directory and see if it works? Just add the name of the test file to the `.env` file and run `make integration`.

**Comment by MaksymMalicki** [comment]
I also see, that your unit tests are not passing. Please run `make unit` and resolve all the issues. Try to take the expected values for the tests, by making prints in the hint itself (the code in the link I sent in the previous comment) and running the modified hint with the prints using `cairo-compile` and `cairo-run` commands. Thanks!

**Comment by MartianGreed** [comment]
Sure, will do

**Comment by MartianGreed** [comment]
> directory

Concerning unit tests, effectively this is not working but since I don't have a math background I tried to solve this by myself but unfortunately didn't succeed. I'm not initiated to modular maths at all...

**Comment by MaksymMalicki** [comment]
> > directory
> 
> Concerning unit tests, effectively this is not working but since I don't have a math background I tried to solve this by myself but unfortunately didn't succeed. I'm not initiated to modular maths at all...

Hey, no need to know maths in depth, just a couple of prints inside of the hint, compiled and run on python cairo-vm will do the job.

File: pkg/hintrunner/zero/zerohint_uint512.go [hunk 7]
```diff
@@ -0,0 +1,90 @@
...
+			if err != nil {
+				return err
+			}
+
+			x := Pack(128, xLoLow, xLoHigh, xHiLow, xHiHigh)
+			p := Pack(128, pLow, pHigh)
+
+			xInverseModPBig := new(big.Int).Exp(&x, big.NewInt(-1), &p)
+
+			split := func(num big.Int, numBitsShift uint16, length int) []fp.Element {
```
Comment (MaksymMalicki) [inline pkg/hintrunner/zero/zerohint_uint512.go:55]: if you are not using this function more than once, then I think there is no need to make it anonymous. Just put the code as it is

File: pkg/hintrunner/zero/zerohint_uint512.go [hunk 8]
```diff
@@ -0,0 +1,87 @@
...
+			x := Pack(BITSHIFT, xLoLow, xLoHigh, xHiLow, xHiHigh)
+			p := Pack(BITSHIFT, pLow, pHigh)
+
+			xInverseModPBig := new(big.Int).Exp(&x, big.NewInt(-1), &p)
+
+			// split big.Int into two fp.Elements
+			xInverseModPSplit := make([]fp.Element, 2)
+			mask := new(big.Int).Sub(new(big.Int).Lsh(big.NewInt(1), uint(BITSHIFT)), big.NewInt(1))
+
+			for i := 0; i < 2; i++ {
```
Comment (TAdev0) [inline pkg/hintrunner/zero/zerohint_uint512.go:61]: it seems there is some optimization to be done here:

do we really need a loop?  we are applying the And operator a second time and then a Rsh operation even though we dont need it.

cant we just do something like 

```
xInverseModPSplit[0] = *new(fp.Element).SetBigInt(new(big.Int).And(xInverseModPBig, mask))
xInverseModPBig.Rsh(xInverseModPBig, uint(BITSHIFT))
xInverseModPSplit[1] = *new(fp.Element).SetBigInt(xInverseModPBig)

```
?

**Review by TAdev0 — CHANGES_REQUESTED** [review]
hi @MartianGreed ,

great work overall!!!
just left one comment for optimization

File: pkg/hintrunner/zero/zerohint_uint512.go (hunk 8 above)
Comment (MartianGreed) [reply pkg/hintrunner/zero/zerohint_uint512.go]: you are totally right ! sorry for that

Comment (TAdev0) [reply pkg/hintrunner/zero/zerohint_uint512.go]: no worries aha

Comment (MartianGreed) [reply pkg/hintrunner/zero/zerohint_uint512.go]: I fixed conflicts !

File: pkg/hintrunner/zero/zerohint_utils.go [hunk 9]
```diff
@@ -78,6 +78,47 @@ func GetUint256ExpandAsFelts(vm *VM.VirtualMachine, ref hinter.ResOperander) ([]
 	return uint256Expanded, nil
 }
 
+func GetUint512AsFelts(vm *VM.VirtualMachine, ref hinter.ResOperander) (*fp.Element, *fp.Element, *fp.Element, *fp.Element, error) {
+	var fps [4]*fp.Element
+	for i := 0; i < 4; i++ {
+		firstRefAddr, err := ref.GetAddress(vm)
```
Comment (TAdev0) [inline pkg/hintrunner/zero/zerohint_utils.go:84]: can you mvoe 

```
	firstRefAddr, err := ref.GetAddress(vm)
		if err != nil {
			return nil, nil, nil, nil, err
		}
```

outside of the loop?

**Review by TAdev0 — COMMENTED** [review]
just left one comment, otherwise great job !

**Review by TAdev0 — APPROVED** [review]
lgtm! great work @MartianGreed

**Pull request merged**